### PR TITLE
[Go SDK] Update beam infra to Go 1.20.2

### PR DESCRIPTION
**Pull request by lostluck**
To ensure the boot loaders are built from the most up to date Go standard library version.

Also strips some needless version specifying in the self-hosted-runner set up. Go doesn't have the same runtime version issues Python and Java have, and should follow the latest Go version. Either the action needs Go or it doesn't.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by lostluck** [comment]
R: @jrmccluskey  as the release manager for 2.47

**Review by jrmccluskey — APPROVED** [review]
LGTM, thank you!

**Comment by lostluck** [comment]
Thanks! Merging once the tests pass.

**Comment by lostluck** [comment]
Run Java PreCommit

**Comment by lostluck** [comment]
Run Java_Kafka_IO_Direct PreCommit

**Comment by lostluck** [comment]
Run Spotless PreCommit

**Comment by lostluck** [comment]
Weird... mongo db failures.

**Comment by lostluck** [comment]
Run CommunityMetrics PreCommit

**Comment by lostluck** [comment]
Run Go PreCommit

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Comment by lostluck** [comment]
Run Java PreCommit

**Comment by lostluck** [comment]
Run Java PreCommit

**Comment by lostluck** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by lostluck** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by lostluck** [comment]
Run Java_Examples_Dataflow_Java17 PreCommit

**Comment by lostluck** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by lostluck** [comment]
Not entirely sure about the mongodb failures. It's like the splitting has changed somehow with the Go version, which is very weird.

https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Phrase/173/

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Comment by lostluck** [comment]
Run Java_Examples_Dataflow_Java17 PreCommit

**Comment by lostluck** [comment]
Run Java_MongoDb_IO_Direct PreCommit

**Comment by lostluck** [comment]
Run Java_PVR_Flink_Docker PreCommit

**Comment by lostluck** [comment]
Run Java_Parquet_IO_Direct PreCommit

**Comment by lostluck** [comment]
Run Java_Spark3_Versions PreCommit

**Comment by lostluck** [comment]
Run Kotlin_Examples PreCommit

**Comment by lostluck** [comment]
Run Portable_Python PreCommit

**Comment by lostluck** [comment]
Run Python PreCommit

**Comment by lostluck** [comment]
Run PythonDocs PreCommit

**Comment by lostluck** [comment]
Run Python_Coverage PreCommit

**Comment by lostluck** [comment]
Run Python_Dataframes PreCommit

**Comment by lostluck** [comment]
Run Python_Examples PreCommit

**Comment by lostluck** [comment]
Run Python_Integration PreCommit

**Comment by lostluck** [comment]
Run Python_PVR_Flink PreCommit

**Comment by lostluck** [comment]
Run Python_Runners PreCommit

**Comment by lostluck** [comment]
Run SQL PreCommit

**Comment by lostluck** [comment]
Run SQL_Java11 PreCommit

**Comment by lostluck** [comment]
Run SQL_Java17 PreCommit

**Comment by lostluck** [comment]
Run Typescript PreCommit

**Comment by lostluck** [comment]
Run Website_Stage_GCS PreCommit

**Comment by lostluck** [comment]
Run Whitespace PreCommit

**Comment by lostluck** [comment]
Run Java_Spark3_Versions PreCommit

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Comment by lostluck** [comment]
Run Go PostCommit

**Comment by lostluck** [comment]
@johannaojeling FYI, I think go 1.20 breaks use of generic coder functions, since the type names no longer keep their new types. This affects the mongodbio types you produced. It's happened due to the various generic/fixes in 1.20, but definitely an unexpected result here.

Raw output is here (search for `--- FAIL`): https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Phrase/189/consoleText

```
	decoding custom coder name:"mongodbio.idRangeRestriction" type:{kind:EXTERNAL external_key:"github.com/apache/beam/sdks/v2/go/pkg/beam/io/mongodbio.idRangeRestriction"} enc:{name:"github.com/apache/beam/sdks/v2/go/pkg/beam/io/mongodbio.encodeBSON[...]" type:{kind:FUNC parameter_types:{kind:EXTERNAL external_key:"github.com/apache/beam/sdks/v2/go/pkg/beam/io/mongodbio.idRangeRestriction"} return_types:{kind:SLICE element:{kind:UINT8}} return_types:{kind:SPECIAL special:ERROR}}} dec:{name:"github.com/apache/beam/sdks/v2/go/pkg/beam/io/mongodbio.decodeBSON[...]" type:{kind:FUNC parameter_types:{kind:SLICE element:{kind:UINT8}} return_types:{kind:EXTERNAL external_key:"github.com/apache/beam/sdks/v2/go/pkg/beam/io/mongodbio.idRangeRestriction"} return_types:{kind:SPECIAL special:ERROR}}}
```

Basically the function/type name ends up as:
`github.com/apache/beam/sdks/v2/go/pkg/beam/io/mongodbio.encodeBSON[...]` for both the idRange and idRangeRestriction encode and decode functions.

I'll send out work around PR which just adds explicit wrapper functions for now. But this is certainly an unexpected side effect of the generic functions on beam's registrations...

**Comment by lostluck** [comment]
https://github.com/apache/beam/issues/26066 filed for the generic coder issue which will be fixed out of band.

**Comment by johannaojeling** [comment]
> @johannaojeling FYI, I think go 1.20 breaks use of generic coder functions, since the type names no longer keep their new types. This affects the mongodbio types you produced. It's happened due to the various generic/fixes in 1.20, but definitely an unexpected result here.

Ah that's interesting. Thanks for identifying the issue and creating a workaround

**Comment by lostluck** [comment]
Run GoPortable PreCommit
